magda-esri-portal-connector/src/EsriPortal.ts

```diff
@@ -0,0 +1,563 @@
+        return undefined;
+    }
+
+    public searchFirstClassOrganizationsByTitle(
+        title: string,
+        maxResults: number
+    ): AsyncPage<any[]> {
+        return undefined;
+    }
+
+    public getJsonDatasetPublisherId(dataset: any): string {
+        return undefined;
+    }
+
+    getJsonDatasetPublisher(dataset: any): Promise<any> {
+        return undefined;
+    }
+
+    private requestGroupInformation(contentId: string): Promise<any> {
+        return new Promise<any>((resolve, reject) => {
+            request(
+                this.urlBuilder.getContentItemGroups(contentId),
+                { json: true },
+                (error, response, body) => {
+                    if (error) {
+                        reject(error);
+                        return;
+                    }
+                    resolve(body);
+                }
+            );
+        });
+    }
+
+    private requestDistributionInformation(
+        distributionUrl: string
+    ): Promise<any> {
+        return new Promise<any>((resolve, reject) => {
+            request(
+                this.urlBuilder.getResource(distributionUrl),
+                { json: true },
+                (error, response, body) => {
+                    if (error) {
+                        reject(error);
+                        return;
+                    }
+                    resolve(body);
+                }
+            );
+        });
+    }
+
+    private requestDataSearchPage(
+        url: uri.URI,
+        startIndex: number
+    ): Promise<EsriPortalDataSearchResponse> {
+        const pageUrl = url.clone();
+        pageUrl.addSearch("start", startIndex);
+        pageUrl.addSearch("num", this.pageSize);
+        const that = this;
+        const operation = async () =>
+            new Promise<EsriPortalDataSearchResponse>((resolve, reject) => {
+                const requestUrl = pageUrl.toString();
+                console.log(
+                    `Requesting start = ${startIndex}, num = ${this.pageSize}`
+                );
+
+                request(
+                    requestUrl,
+                    { json: true },
+                    async (error, response, body) => {
+                        if (error) {
+                            reject(error);
+                            return;
+                        }
+
+                        // A single portal item only has one distribution in the form of a url.
+                        // That url however may represent a single layer, or a map (containing multiple layers)
+                        // The bulk of the distribution information needs to be retrieved
+                        // from the map or feature service endpoint
+                        // An individual item may result in one or many additional requests
+                        for (let i = 0; i < body.results.length; ++i) {
+                            const item = body.results[i];
```

Comment:
This is maybe a question of style, but this is a really long loop and I find it a bit hard to follow the flow. Is there any chance of breaking this up into a few more functions and getting rid of the `continue`s?

So it's more like
```javascript
if (isShared(item)) {
    return processSharedItem(item);
} else if (isPublic(item)) {
    return processPublicItem(item);
} else if (isPrivate(item)) {
   return processPrivateItem(item);
} else {
 // etc
}
```